EX1/HW7.py

```diff
@@ -0,0 +1,373 @@
+
+    :param list usernames: Users for which we should create PO Boxes.
+    """
+
+    def __init__(self, usernames):
+        self.message_id = 0
+        self.boxes = {user: [] for user in usernames}
+
+    def send_message(self, sender, recipient, message_body, urgent=False):
+        """Send a message to a recipient.
+
+        :param str sender: The message sender's username.
+        :param str recipient: The message recipient's username.
+        :param str message_body: The body of the message.
+        :param urgent: The urgency of the message.
+        :type urgent: bool, optional
+        :return: The message ID, auto incremented number.
+        :rtype: int
+        :raises KeyError: if the recipient does not exist.
+        """
+        user_box = self.boxes[recipient]
+        self.message_id += 1
+        message_details = {
+            'id': self.message_id,
+            'body': message_body,
+            'sender': sender,
+            'read': False  # Added to track whether message has been read
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, username, num_msgs=None):
+        """Read a number of messages in a user's inbox.
+
+        :param str username: The name of the user whose inbox we want to read.
+        :param int num_msgs: Optional, the number of messages to read. Default is None, which will read all messages.
+        :return: A list of dictionaries representing the messages read.
+        :rtype: list[dict]
+        """
+        user_box = self.boxes[username]
+        if num_msgs is None:
+            num_msgs = len(user_box)
+        msgs = []
+        for msg in user_box:
+            if not msg['read']:
+                msgs.append(msg)
+                msg['read'] = True
+            if len(msgs) == num_msgs:
+                break
+        return msgs
+
+    def search_inbox(self, username, search_string):
+        """Search for messages in a user's inbox containing a search string.
+
+        :param str username: The name of the user whose inbox we want to search.
+        :param str search_string: The string to search for.
+        :return: A list of dictionaries representing the messages containing the search string.
+        :rtype: list[dict]
+        """
+        user_box = self.boxes[username]
+        found_msgs = []
+        for msg in user_box:
+            if search_string in msg['body']:
+                found_msgs.append(msg)
+        return found_msgs
+
+
+class Player:
+    def __init__(self, name):
+        self.name = name
+        self.hp = 100
+        self.exp = 0
+        self.level = 1
+        self.nemeses = []
+        self.attacked_by = []
+
+    def attack(self, enemy_name=None):
+        if enemy_name is None:
+            enemy = self.nemeses[-1]
+        else:
+            for enemy in self.nemeses:
```

Comment:
You should aim to write everything in a single line, even if you can. 
This link https://towardsdatascience.com/you-should-stop-craving-for-python-one-liners-5dbf8e0de23d explains it well, even though I disagree with how important memory and runtime are in most python programs.